From: SeongJae Park <sj@kernel.org>
Date: Thu, 8 Jan 2026 18:28:48 -0800
Subject: [PATCH] selftests/damon: ensure access_memory run until test finishes

Depending on testing system speed, access_memory might finish earlier
than DAMON tests.  Ensure it runs until test is done.

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 tools/testing/selftests/damon/access_memory.c        | 12 +++++++-----
 .../testing/selftests/damon/damos_apply_interval.py  |  5 +++++
 tools/testing/selftests/damon/damos_quota.py         |  3 +++
 tools/testing/selftests/damon/damos_quota_goal.py    |  4 ++--
 ...fs_update_schemes_tried_regions_wss_estimation.py |  3 +++
 5 files changed, 20 insertions(+), 7 deletions(-)

diff --git a/tools/testing/selftests/damon/access_memory.c b/tools/testing/selftests/damon/access_memory.c
index 56b17e8fe1be..035fb51143e0 100644
--- a/tools/testing/selftests/damon/access_memory.c
+++ b/tools/testing/selftests/damon/access_memory.c
@@ -31,11 +31,13 @@ int main(int argc, char *argv[])
 	for (i = 0; i < nr_regions; i++)
 		regions[i] = malloc(sz_region);
 
-	for (i = 0; i < nr_regions; i++) {
-		start_clock = clock();
-		while ((clock() - start_clock) * 1000 / CLOCKS_PER_SEC <
-				access_time_ms)
-			memset(regions[i], i, sz_region);
+	while (1) {
+		for (i = 0; i < nr_regions; i++) {
+			start_clock = clock();
+			while ((clock() - start_clock) * 1000 / CLOCKS_PER_SEC
+					< access_time_ms)
+				memset(regions[i], i, sz_region);
+		}
 	}
 	return 0;
 }
diff --git a/tools/testing/selftests/damon/damos_apply_interval.py b/tools/testing/selftests/damon/damos_apply_interval.py
index f04d43702481..b1782f245b14 100755
--- a/tools/testing/selftests/damon/damos_apply_interval.py
+++ b/tools/testing/selftests/damon/damos_apply_interval.py
@@ -41,12 +41,17 @@ def main():
 
     wss_collected = []
     nr_quota_exceeds = 0
+    nr_tried = 0
     while proc.poll() == None:
         time.sleep(0.1)
         err = kdamonds.kdamonds[0].update_schemes_stats()
         if err != None:
             print('stats update failed: %s' % err)
             exit(1)
+        nr_tried += 1
+        if nr_tried == 40:
+            break
+    proc.terminate()
     schemes = kdamonds.kdamonds[0].contexts[0].schemes
     nr_tried_stats = [s.stats.nr_tried for s in schemes]
     if nr_tried_stats[0] == 0 or nr_tried_stats[1] == 0:
diff --git a/tools/testing/selftests/damon/damos_quota.py b/tools/testing/selftests/damon/damos_quota.py
index 57c4937aaed2..758936555f56 100755
--- a/tools/testing/selftests/damon/damos_quota.py
+++ b/tools/testing/selftests/damon/damos_quota.py
@@ -49,6 +49,9 @@ def main():
         scheme = kdamonds.kdamonds[0].contexts[0].schemes[0]
         wss_collected.append(scheme.tried_bytes)
         nr_quota_exceeds = scheme.stats.qt_exceeds
+        if len(wss_collected) == 40:
+            break
+    proc.terminate()
 
     wss_collected.sort()
     nr_expected_quota_exceeds = 0
diff --git a/tools/testing/selftests/damon/damos_quota_goal.py b/tools/testing/selftests/damon/damos_quota_goal.py
index f76e0412b564..960f89e26a8c 100755
--- a/tools/testing/selftests/damon/damos_quota_goal.py
+++ b/tools/testing/selftests/damon/damos_quota_goal.py
@@ -33,8 +33,7 @@ def main():
     score_values_to_test = [0, 15000, 5000, 18000]
     while proc.poll() == None:
         if len(score_values_to_test) == 0:
-            time.sleep(0.1)
-            continue
+            break
 
         goal.current_value = score_values_to_test.pop(0)
         expect_increase = goal.current_value < goal.target_value
@@ -75,6 +74,7 @@ def main():
                   (expect_increase, increased))
             exit(1)
         last_effective_bytes = goal.effective_bytes
+    proc.terminate()
 
 if __name__ == '__main__':
     main()
diff --git a/tools/testing/selftests/damon/sysfs_update_schemes_tried_regions_wss_estimation.py b/tools/testing/selftests/damon/sysfs_update_schemes_tried_regions_wss_estimation.py
index a000982c59be..206bd4f312f2 100755
--- a/tools/testing/selftests/damon/sysfs_update_schemes_tried_regions_wss_estimation.py
+++ b/tools/testing/selftests/damon/sysfs_update_schemes_tried_regions_wss_estimation.py
@@ -36,6 +36,9 @@ def main():
 
         wss_collected.append(
                 kdamonds.kdamonds[0].contexts[0].schemes[0].tried_bytes)
+        if len(wss_collected) == 40:
+            break
+    proc.terminate()
 
     wss_collected.sort()
     percentile = 75
-- 
2.47.3

